libbeat/outputs/kafka/partition.go

```diff
@@ -0,0 +1,297 @@
+package kafka
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash"
+	"hash/fnv"
+	"math/rand"
+	"strconv"
+
+	"github.com/Shopify/sarama"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+type partitionBuilder func(*common.Config) (func() partitioner, error)
+
+type partitioner func(*message, int32) (int32, error)
+
+// stablePartitioner re-uses last configured partition in case of event being
+// repartitioned (on retry from libbeat).
+type messagePartitioner struct {
+	p          partitioner
+	reachable  bool
+	partitions int32 // number of partitions seen last
+}
+
+func makePartitioner(
+	partition map[string]*common.Config,
+) (sarama.PartitionerConstructor, error) {
+	mkStrategy, reachable, err := initPartitionStrategy(partition)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(topic string) sarama.Partitioner {
+		return &messagePartitioner{
+			p:         mkStrategy(),
+			reachable: reachable,
+		}
+	}, nil
+}
+
+var partitioners = map[string]partitionBuilder{
+	"random":      cfgRandomPartitioner,
+	"round_robin": cfgRoundRobinPartitioner,
+	"hash":        cfgHashPartitioner,
+}
+
+func initPartitionStrategy(
+	partition map[string]*common.Config,
```

Comment:
partitioners are configured by name. the name in map[string] is for lookup into `partitioners`. We use this pattern in a few places with `plugins` being given by name. Alternate solution: capture *common.Config, get list of keys and check it's length is 1, extract Child config by key.